In [1]:
# import libraries
import numpy as np
import pandas as pd
from sklearn import linear_model
import matplotlib.pyplot as plt
import statsmodels.api as sm
import seaborn as sns

# 1. Weather model

1. First, load the dataset from the weatherinszeged table from Thinkful's database.
2. Like in the previous checkpoint, build a linear regression model where your target variable is the difference between the apparenttemperature and the temperature. As explanatory variables, use humidity and windspeed. Now, estimate your model using OLS. What are the R-squared and adjusted R-squared values? Do you think they are satisfactory? Why?
3. Next, include the interaction of humidity and windspeed to the model above and estimate the model using OLS. Now, what is the R-squared of this model? Does this model improve upon the previous one?
4. Add visibility as an additional explanatory variable to the first model and estimate it. Did R-squared increase? What about adjusted R-squared? Compare the differences put on the table by the interaction term and the visibility in terms of the improvement in the adjusted R-squared. Which one is more useful?
5. Choose the best one from the three models above with respect to their AIC and BIC scores. Validate your choice by discussing your justification with your mentor.

In [2]:
# import dataset
df_w = pd.read_csv(r'C:\Users\AP\Desktop/weatherHistory.csv')

# columns to lower case
df_w.columns = map(str.lower, df_w.columns)

# rename columns/fix typos
df_w.columns = ['date', 'summary', 'precip type', 'temp',
       'apparent temp', 'humidity', 'windspeed',
       'windbearing', 'visibility', 'cloud cover',
       'pressure', 'daily summary']

In [4]:
# Y is the target variable
Y = df_w['apparent temp'] - df_w['temp']
# X is the feature set
X = df_w[['humidity','windspeed']]

X = sm.add_constant(X)

results = sm.OLS(Y, X).fit()

results.summary()

c:\users\ap\appdata\local\programs\python\python37-32\lib\site-packages\numpy\core\fromnumeric.py:2389: FutureWarning: Method .ptp is deprecated and will be removed in a future version. Use numpy.ptp instead.
  return ptp(axis=axis, out=out, **kwargs)


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.288
Model:                            OLS   Adj. R-squared:                  0.288
Method:                 Least Squares   F-statistic:                 1.949e+04
Date:                Tue, 24 Sep 2019   Prob (F-statistic):               0.00
Time:                        21:05:24   Log-Likelihood:            -1.7046e+05
No. Observations:               96453   AIC:                         3.409e+05
Df Residuals:                   96450   BIC:                         3.409e+05
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          2.4381      0.021    115.948      0.000       2.397       2.479
humidity      -3.0292      0.024   -126.479      0.000      -3.076      -2.982
windspeed     -0.1193      0.001   -176.164      0.000      -0.121      -0.118
==============================================================================
Omnibus:                     3935.747   Durbin-Watson:                   0.264
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4613.311
Skew:                          -0.478   Prob(JB):                         0.00
Kurtosis:                       3.484   Cond. No.                         88.1
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

The R-squared and adjusted R-squared are even at 0.288. The range is from 0-1 so this score seems low. This means that 29% of the target variable can be explained by the features in the model, which also means 71% is unexplained variance. 

In [7]:
# interaction between humidity and windspeed (multiplication)
df_w['humidity_windspeed_interaction'] = df_w.humidity * df_w.windspeed

# Y is the target variable
Y = df_w['apparent temp'] - df_w['temp']
# X is the feature set
X = df_w[['humidity','windspeed', 'humidity_windspeed_interaction']]

X = sm.add_constant(X)

results = sm.OLS(Y, X).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.341
Model:                            OLS   Adj. R-squared:                  0.341
Method:                 Least Squares   F-statistic:                 1.666e+04
Date:                Tue, 24 Sep 2019   Prob (F-statistic):               0.00
Time:                        21:14:49   Log-Likelihood:            -1.6669e+05
No. Observations:               96453   AIC:                         3.334e+05
Df Residuals:                   96449   BIC:                         3.334e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==================================================================================================
                                     coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------------------
const                              0.0839      0.033      2.511      0.012       0.018       0.149
humidity                           0.1775      0.043      4.133      0.000       0.093       0.262
windspeed                          0.0905      0.002     36.797      0.000       0.086       0.095
humidity_windspeed_interaction    -0.2971      0.003    -88.470      0.000      -0.304      -0.291
==============================================================================
Omnibus:                     4849.937   Durbin-Watson:                   0.262
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             9295.404
Skew:                          -0.378   Prob(JB):                         0.00
Kurtosis:                       4.320   Cond. No.                         193.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Both R-squared and adjusted R-squared increased, but are still equal, at 0.341. This is an improvement on the previous version of the model. 

In [9]:
# interaction between humidity and windspeed (multiplication)
df_w['humidity_windspeed_interaction'] = df_w.humidity * df_w.windspeed

# Y is the target variable
Y = df_w['apparent temp'] - df_w['temp']
# X is the feature set
X = df_w[['humidity','windspeed', 'visibility']]

X = sm.add_constant(X)

results = sm.OLS(Y, X).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.304
Model:                            OLS   Adj. R-squared:                  0.303
Method:                 Least Squares   F-statistic:                 1.401e+04
Date:                Tue, 24 Sep 2019   Prob (F-statistic):               0.00
Time:                        21:19:18   Log-Likelihood:            -1.6938e+05
No. Observations:               96453   AIC:                         3.388e+05
Df Residuals:                   96449   BIC:                         3.388e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          1.5756      0.028     56.605      0.000       1.521       1.630
humidity      -2.6066      0.025   -102.784      0.000      -2.656      -2.557
windspeed     -0.1199      0.001   -179.014      0.000      -0.121      -0.119
visibility     0.0540      0.001     46.614      0.000       0.052       0.056
==============================================================================
Omnibus:                     3833.895   Durbin-Watson:                   0.279
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4584.022
Skew:                          -0.459   Prob(JB):                         0.00
Kurtosis:                       3.545   Cond. No.                         131.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

After adding visibility to the feature set, both R-squared and adjusted R-squared increased slightly to about equal at 0.304 and 0.303, respectively. This is a meager improvement from compared to the second model using the interaction between windspeed and humidity. Looking at the R-squared and adjusted R-squared values, the second model performed the best and explained the most variance of the three models.

According to the AIC and BIC numbers, the second model performs the best because both AIC and BIC are the lowest between the three.

Taking into account both the adjusted R-squared and AIC/BIC numbers, the second model performs the best out of the three.

--------------------



# 2. House prices model

1. Load the houseprices data from Thinkful's database.
2. Run your house prices model again and assess the goodness of fit of your model using F-test, R-squared, adjusted R-squared, AIC and BIC.
3. Do you think your model is satisfactory? If so, why?
4. In order to improve the goodness of fit of your model, try different model specifications by adding or removing some variables.
5. For each model you try, get the goodness of fit metrics and compare your models with each other. Which model is the best and why?

In [3]:
# import dataset
df = pd.read_csv(r'C:\Users\AP\Downloads\house prices.csv')

# columns to lower case
df.columns = map(str.lower, df.columns)

# list/number of numerical columns
num_col = df.select_dtypes(['int64', 'float64']).columns

# list/number of alpha columns
alpha_col = df.select_dtypes(['object']).columns

# convert alpha features to binary
df['mszoning'] = pd.get_dummies(df.mszoning, drop_first=True)
df['neighborhood'] = pd.get_dummies(df.neighborhood, drop_first=True)
df['exterior1st'] = pd.get_dummies(df.exterior1st, drop_first=True)

In [10]:
# OLS parameters
# define variables for model
X = df[['overallqual', 'grlivarea', 'garagecars', 'mszoning', 'neighborhood', 'exterior1st']]
Y = df.saleprice

# create constant
X = sm.add_constant(X)

results = sm.OLS(Y, X).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              saleprice   R-squared:                       0.740
Model:                            OLS   Adj. R-squared:                  0.739
Method:                 Least Squares   F-statistic:                     688.8
Date:                Tue, 24 Sep 2019   Prob (F-statistic):               0.00
Time:                        21:34:34   Log-Likelihood:                -17561.
No. Observations:                1460   AIC:                         3.514e+04
Df Residuals:                    1453   BIC:                         3.517e+04
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
================================================================================
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
const        -9.953e+04   4872.194    -20.429      0.000   -1.09e+05      -9e+04
overallqual   2.738e+04   1087.034     25.185      0.000    2.52e+04    2.95e+04
grlivarea       50.2908      2.565     19.607      0.000      45.259      55.322
garagecars    2.133e+04   1810.777     11.778      0.000    1.78e+04    2.49e+04
mszoning     -8029.3049   5255.941     -1.528      0.127   -1.83e+04    2280.738
neighborhood -3.991e+04   2.87e+04     -1.389      0.165   -9.63e+04    1.65e+04
exterior1st  -2.103e+04   4.07e+04     -0.516      0.606   -1.01e+05    5.89e+04
==============================================================================
Omnibus:                      415.739   Durbin-Watson:                   1.991
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             9565.906
Skew:                           0.771   Prob(JB):                         0.00
Kurtosis:                      15.445   Cond. No.                     6.15e+04
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 6.15e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

R-squared and adjusted R-squared are 0.74 and 0.739, respectively. 74% of the variance in the target variable is explained by the model, and 26% is unexplained variance. I do not believe this is satisfactory. Let's try and capture more of the unexplained variance by doing a few things by creating and adjusting some features.

* creating a new variable accounting for total square footage of the house
* create a new interaction variable between the new totalsf variable and overallqual
* log-transform the target variable

In [18]:
# create totalsf variable
df['totalsf'] = df['totalbsmtsf'] + df['1stflrsf'] + df['2ndflrsf']

# Y is the target variable
Y = df['saleprice']
# X is the feature set
X = df[['overallqual', 'grlivarea', 'garagecars', 'mszoning', 'neighborhood', 'exterior1st', 'totalsf']]

X = sm.add_constant(X)

results = sm.OLS(Y, X).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              saleprice   R-squared:                       0.762
Model:                            OLS   Adj. R-squared:                  0.761
Method:                 Least Squares   F-statistic:                     663.0
Date:                Tue, 24 Sep 2019   Prob (F-statistic):               0.00
Time:                        21:50:18   Log-Likelihood:                -17497.
No. Observations:                1460   AIC:                         3.501e+04
Df Residuals:                    1452   BIC:                         3.505e+04
Df Model:                           7                                         
Covariance Type:            nonrobust                                         
================================================================================
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
const        -9.897e+04   4665.263    -21.215      0.000   -1.08e+05   -8.98e+04
overallqual   2.345e+04   1095.072     21.415      0.000    2.13e+04    2.56e+04
grlivarea       13.7720      4.008      3.436      0.001       5.910      21.634
garagecars    1.827e+04   1753.995     10.414      0.000    1.48e+04    2.17e+04
mszoning       105.7510   5081.664      0.021      0.983   -9862.437    1.01e+04
neighborhood -2.619e+04   2.75e+04     -0.951      0.342   -8.02e+04    2.78e+04
exterior1st   4521.9197   3.91e+04      0.116      0.908   -7.21e+04    8.11e+04
totalsf         32.6197      2.829     11.529      0.000      27.070      38.170
==============================================================================
Omnibus:                      431.349   Durbin-Watson:                   1.972
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            38617.200
Skew:                          -0.318   Prob(JB):                         0.00
Kurtosis:                      28.187   Cond. No.                     1.20e+05
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.2e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [21]:
# Y is the target variable (log transformed)
Y = np.log1p(df['saleprice'])
# X is the feature set
X = df[['overallqual', 'grlivarea', 'garagecars', 'mszoning', 'neighborhood', 'exterior1st', 'totalsf', 'int_over_sf']]

X = sm.add_constant(X)

results = sm.OLS(Y, X).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              saleprice   R-squared:                       0.805
Model:                            OLS   Adj. R-squared:                  0.804
Method:                 Least Squares   F-statistic:                     749.5
Date:                Tue, 24 Sep 2019   Prob (F-statistic):               0.00
Time:                        21:53:37   Log-Likelihood:                 462.59
No. Observations:                1460   AIC:                            -907.2
Df Residuals:                    1451   BIC:                            -859.6
Df Model:                           8                                         
Covariance Type:            nonrobust                                         
================================================================================
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
const           10.1762      0.053    191.672      0.000      10.072      10.280
overallqual      0.1861      0.009     21.206      0.000       0.169       0.203
grlivarea     4.911e-05   1.83e-05      2.691      0.007    1.33e-05    8.49e-05
garagecars       0.1127      0.008     14.093      0.000       0.097       0.128
mszoning         0.0322      0.023      1.393      0.164      -0.013       0.078
neighborhood    -0.1259      0.125     -1.005      0.315      -0.372       0.120
exterior1st     -0.0154      0.178     -0.087      0.931      -0.364       0.333
totalsf          0.0003   2.47e-05     13.144      0.000       0.000       0.000
int_over_sf  -2.416e-05   2.88e-06     -8.398      0.000   -2.98e-05   -1.85e-05
==============================================================================
Omnibus:                      503.633   Durbin-Watson:                   1.981
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             3598.933
Skew:                          -1.419   Prob(JB):                         0.00
Kurtosis:                      10.149   Cond. No.                     7.22e+05
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 7.22e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""